Update ui restore page

JustinBeBoy opened the pull request:
Resolve #376
- Update layout in restore seed page
- Custom UI Editor
- Custom UI dropdown
- Custom space vertical

itswisdomagain commented on nuklear/pagehandlers/receive.go:
```diff
@@ -25,30 +25,47 @@ type ReceiveHandler struct {
-		})
+		} else {
+			handler.accountNumber = handler.accountSelectorWidget.GetSelectedAccountNumber()
+			handler.generatedAddress, handler.generateAddressError = handler.wallet.ReceiveAddress(handler.accountNumber)
```
Should not call `handler.wallet.ReceiveAddress` directly in the `Render` method. Reason is `Render` is invoked multiple times per second. I believe this if statement should rather go into `BeforeRender`.

Then you'd need to figure out a way to listen for dropdown selection change event on the account selection widget, and place the following code in the selection change event listener:
```
handler.accountNumber = handler.accountSelectorWidget.GetSelectedAccountNumber()
handler.generatedAddress, handler.generateAddressError = handler.wallet.ReceiveAddress(handler.accountNumber)
window.Master().Changed()
```
```

itswisdomagain commented on nuklear/widgets/tablecell.go:
```diff
@@ -106,3 +106,29 @@ func (link *LinkTableCell) Render(window *Window) {
+
+func (link *LinkLabelCell) MinWidth(window *Window) int {
+	return window.LabelWidth(link.text)
+}
```
Add an extra line at the end of this file

itswisdomagain commented on terminal/pages/receive.go:
```diff
@@ -92,7 +92,7 @@ func receivePage(wallet walletcore.Wallet, hintTextView *primitives.TextView, se
 			}
 		})
-		formButton.AddButton("NEW ADDRESS.", func() {
+		formButton.AddButton("new address.", func() {
```
Please remove these changes that obviously do not relate to this PR. Create a new PR if these changes are important.

Sirmorrison replied:
its was just to change the casing from all caps to lower casing. Too small for a PR.

itswisdomagain replied:
why are you changing it?

Sirmorrison replied:
requested.

itswisdomagain commented on go.mod:
```diff
@@ -2,8 +2,6 @@ module github.com/raedahgroup/godcr
 
 go 1.12
 
-require github.com/raedahgroup/godcr/cmd/godcr-fyne v0.0.0-00010101000000-000000000000 // indirect
```
There shouldn't be any changes to the `go.mod` and `go.sum` files in the project root dir. I suspect this change would cause a build error when trying to build fyne from the project root using `go build ./cmd/godcr-fyne`.

Currently, fyne doesn't build because of a dcrlibwallet mod version issue. That will soon be resolved on master and you'd need to rebase this PR. Then we can properly test the above assertion.

itswisdomagain commented on nuklear/pagehandlers/receive.go:
```diff
@@ -25,30 +25,47 @@ type ReceiveHandler struct {
+		contentWindow.Row(widgets.EditorHeight).Static(contentWindow.LabelWidth("You can also manually generate a"), 120)
+		contentWindow.AddLabelsToCurrentRow(widgets.NewColoredLabelTableCell("You can also manually generate a", widgets.LeftCenterAlign, styles.GrayColor))
+		contentWindow.AddLinkLabelsToCurrentRow(widgets.NewLinkLabelCellCell("new address.", func() {
+			handler.generatedAddress, handler.generateAddressError = handler.wallet.GenerateNewAddress(handler.accountNumber)
```
`handler.accountNumber` is only set once, in `BeforeRender` when the page is navigated to - this initial value is always the account number of the first account in the wallet. This value does not change when the account dropdown selection changes, so clicking this label to generate a new address never uses the current selection in the dropdown.

itswisdomagain commented on nuklear/pagehandlers/receive.go:
```diff
@@ -25,30 +25,48 @@ type ReceiveHandler struct {
 	handler.generatedAddress = ""
+
+	handler.accountSelectorWidget = widgets.AccountSelectorWidget("Account:", false, false, wallet, nil)
+	handler.accounts, handler.generateAddressError = wallet.AccountsOverview(walletcore.DefaultRequiredConfirmations)
```
`widgets.AccountSelectorWidget` already calls `wallet.AccountsOverview`. I'm not okay with having that function called twice: here (line 38) and in the `AccountSelectorWidget`. Find a way to merge them. Specifically, why do we need to have information of the accounts here?

itswisdomagain commented on nuklear/pagehandlers/receive.go:
```diff
@@ -25,30 +25,48 @@ type ReceiveHandler struct {
+			handler.accountNumber = handler.accountSelectorWidget.GetSelectedAccountNumber()
+			handler.generatedAddress, handler.generateAddressError = handler.wallet.ReceiveAddress(handler.accountNumber)
+		})
+	}
```
Modify lines 40-48 thus:

Note the following:
- No need for `handler.accountNumber` or `handler.accounts`.
- The selection changed func no longer needs an int parameter.

```
// first setup account selector widget
handler.accountSelectorWidget = widgets.AccountSelectorWidget("Account:", false, false, wallet, func() {
	handler.generatedAddress, handler.generateAddressError = handler.wallet.ReceiveAddress(handler.accountSelectorWidget.GetSelectedAccountNumber())
	refreshWindowDisplay()
})

// generate address for first account in wallet
handler.generatedAddress, handler.generateAddressError = handler.wallet.ReceiveAddress(handler.accountSelectorWidget.GetSelectedAccountNumber())
```

itswisdomagain commented on nuklear/pagehandlers/receive.go:
```diff
@@ -25,30 +25,48 @@ type ReceiveHandler struct {
+		contentWindow.AddLabelsToCurrentRow(widgets.NewColoredLabelTableCell("You can also manually generate a", widgets.LeftCenterAlign, styles.GrayColor))
+		contentWindow.AddLinkLabelsToCurrentRow(widgets.NewLinkLabelCellCell("new address.", func() {
+			handler.generatedAddress, handler.generateAddressError = handler.wallet.GenerateNewAddress(handler.accountNumber)
+			window.Master().Changed()
```
Other pages, iirc, use a `handler.refreshDisplay()` function. That is probably better.

itswisdomagain commented on nuklear/pagehandlers/receive.go:
```diff
@@ -25,30 +25,48 @@ type ReceiveHandler struct {
-			handler.generatedAddress, handler.generateAddressError = handler.wallet.ReceiveAddress(accountNumber)
-			window.Master().Changed()
-		})
+		window.Master().Changed()
```
Calling master changed will only cause this Render function to be called again. When it is called again, this master changed function will also be re-called and you get an endless cycle of repainting the window. Please remove this.

itswisdomagain commented on nuklear/widgets/tablecell.go:
```diff
@@ -106,3 +106,29 @@ func (link *LinkTableCell) Render(window *Window) {
 	return window.LabelWidth(link.text)
 }
+
+type LinkLabelCell struct {
```
Is there a `label.go` file in this package? This should be there not here, this has no relation with a table. It could preferably be in a `linklabel.go` file instead.

Sirmorrison commented on go.sum:
```diff
@@ -281,6 +281,7 @@ github.com/raedahgroup/dcrlibwallet v1.1.1-0.20190904121310-90259892e503 h1:XEV7
 github.com/raedahgroup/dcrlibwallet v1.1.1-0.20190904121310-90259892e503/go.mod h1:szxgNhCQ/GRyFpmkfYi3T3ZkXPx9Hm9j0EnjaBfJK/o=
 github.com/raedahgroup/dcrlibwallet v1.1.1-0.20190911210329-068758c38d77 h1:cGSRi6zwgpXGmCYCAkEDgX6XdO4Y4MwdFa7LCKDzQ/8=
 github.com/raedahgroup/dcrlibwallet v1.1.1-0.20190911210329-068758c38d77/go.mod h1:6AGgglq9BB0twJnZnHvfH8hIQrbw6yBJSPp9QI+hBk0=
+github.com/raedahgroup/dcrlibwallet v1.1.1-0.20190917220221-47d626dfbe08/go.mod h1:6AGgglq9BB0twJnZnHvfH8hIQrbw6yBJSPp9QI+hBk0=
```
I didn't change it. I just build and the app did its thing.

itswisdomagain commented on nuklear/widgets/accountselector.go:
```diff
@@ -2,6 +2,7 @@ package widgets
 
 import (
 	"github.com/raedahgroup/godcr/app/walletcore"
+	"github.com/aarzilli/nucular/label"
```
Is this import used? If so, need to order it properly, but I can't find where it is used.

Sirmorrison replied:
used inlines 94 and 97.

JustinBeBoy committed:
Custom main layout restore
- custom card layout
- custom restore button

JustinBeBoy committed:
Custom seed editor
- custom seed dropdown
- update list layout
- add action restore

Sirmorrison commented on ui/decredmaterial/restoreeditor.go:
```diff
@@ -0,0 +1,64 @@
+	"image/color"
+
+	"gioui.org/layout"
+	"gioui.org/unit"
```
No need for this. Declare the unit parameter in the dimensions.go file if it does not exist and the use it in your code.

oshorefueled replied:
> No need for this. Declare the unit parameter in the dimensions.go file if it does not exist and the use it in your code.

Most of the code in decredmaterial don't use dimensions defined in the values package. And the reason for that was to make decredmaterial an extension of gio material that is independent of godcr code. That being said, not using the values declared in the values package is quite fine here.

Sirmorrison commented on ui/decredmaterial/restoreeditor.go:
```diff
@@ -0,0 +1,64 @@
+		TitleLabel: t.Body2(title),
+		LineColor:  t.Color.Gray1,
+		m5:         unit.Dp(5),
+		m10:        unit.Dp(10),
```
```suggestion
		m5:        values.MarginPadding5,
		m10:        values.MarginPadding5,
```

oshorefueled replied:
> 

This change isn't required, the reason is stated in the comment above.

Sirmorrison commented on ui/decredmaterial/restoreeditor.go:
```diff
@@ -0,0 +1,64 @@
+}
+
+func (re RestoreEditor) Layout(gtx layout.Context) layout.Dimensions {
+	border := widget.Border{Color: re.LineColor, CornerRadius: unit.Dp(8), Width: unit.Dp(2)}
```
```suggestion
	border := widget.Border{Color: re.LineColor, CornerRadius: values.MarginPadding8, Width: values.MarginPadding2}
```

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -532,7 +637,43 @@ func (pg *createRestore) editorSeedsEventsHandler() {
+
+	inset := layout.Inset{
+		Top:  unit.Dp(35),
+		Left: unit.Dp(0),
```
```suggestion
		Top:  values.MarginPadding35,
		Left: values.MarginPadding0,
```

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -532,7 +637,43 @@ func (pg *createRestore) editorSeedsEventsHandler() {
+
+	m := op.Record(gtx.Ops)
+	inset.Layout(gtx, func(gtx C) D {
+		border := widget.Border{Color: pg.theme.Color.LightGray, CornerRadius: unit.Dp(5), Width: unit.Dp(2)}
```
```suggestion
		border := widget.Border{Color: pg.theme.Color.LightGray, CornerRadius: values.MarginPadding5, Width: values.MarginPadding2}
```

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -532,7 +637,43 @@ func (pg *createRestore) editorSeedsEventsHandler() {
+			return pg.optionsMenuCard.Layout(gtx, func(gtx C) D {
+				return (&layout.List{Axis: layout.Vertical}).Layout(gtx, len(pg.seedMenu), func(gtx C, i int) D {
+					return material.Clickable(gtx, pg.seedMenu[i].button, func(gtx C) D {
+						return layout.UniformInset(unit.Dp(10)).Layout(gtx, func(gtx C) D {
```
```suggestion
						return layout.UniformInset(values.MarginPadding10).Layout(gtx, func(gtx C) D {
```

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -160,20 +178,22 @@ func (pg *createRestore) layout(gtx layout.Context, common pageCommon) layout.Di
-				})
+			layout.Rigid(func(gtx C) D {
+				// return layout.Flex{Axis: layout.Vertical}.Layout(gtx,
+				// 	layout.Rigid(func(gtx C) D {
```
Removed the commented codes if they are no longer needed.

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -160,20 +178,22 @@ func (pg *createRestore) layout(gtx layout.Context, common pageCommon) layout.Di
+					})
+				}
+				// 	}),
+				// )
```
Removed the commented codes if they are no longer needed.

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -323,64 +307,222 @@ func (pg *createRestore) mainContent(gtx layout.Context) layout.Dimensions {
+					})
+				}),
+				layout.Rigid(func(gtx C) D {
+					return common.UniformPadding(gtx, func(gtx C) D {
```
Currently common.UniformPadding() uses a uniform inset of  24. This should be the reason you have too much padding between you page title and the seed phrase section which should be 16 Dp.

My recommendation is to manually define the Inset values instead of using common.UniformPadding().
```suggestion
m := values.MarginPadding24
		return layout.Inset{
		        Top: values.MarginPadding24,
			Left: m,
			Right: m,
			Bottom: m,
		}.Layout(gtx, func(gtx C) D {
```

<img width="532" alt="image" src="https://user-images.githubusercontent.com/27733432/116500358-50e2a400-a8a6-11eb-88b5-d73bb7289d9a.png">

too much padding. it should be 16. Kindly
<img width="473" alt="image" src="https://user-images.githubusercontent.com/27733432/116500416-6f489f80-a8a6-11eb-9be1-59cf840db65d.png">

oshorefueled replied:
The Container type can be used. It is a wrapper for the inset type that was added to differentiate when an inset is used as a margin or padding in the code.

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -323,64 +307,222 @@ func (pg *createRestore) mainContent(gtx layout.Context) layout.Dimensions {
+										CornerRadius: values.MarginPadding14,
+										Width:        values.MarginPadding2,
+									}
+									phase := pg.theme.Body1(phaseProgress)
```
```suggestion
									phase := pg.theme.Body2(phaseProgress)
```

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -323,64 +307,222 @@ func (pg *createRestore) mainContent(gtx layout.Context) layout.Dimensions {
+									border := widget.Border{
+										Color:        pg.theme.Color.Gray1,
+										CornerRadius: values.MarginPadding14,
+										Width:        values.MarginPadding2,
```
```suggestion
										Width:        values.MarginPadding1,
```

Sirmorrison commented on ui/create_restore_page.go:
```diff
@@ -323,64 +307,222 @@ func (pg *createRestore) mainContent(gtx layout.Context) layout.Dimensions {
+	return pg.walletName.Layout(gtx)
+}
+
+func (pg *createRestore) restorePageSections(gtx layout.Context, title string, phaseProgress string, body layout.Widget) layout.Dimensions {
```
<img width="391" alt="image" src="https://user-images.githubusercontent.com/27733432/116502987-e5500500-a8ac-11eb-8047-7eb725163261.png">

<img width="323" alt="image" src="https://user-images.githubusercontent.com/27733432/116503058-14667680-a8ad-11eb-8527-83932d8f51db.png">

Too much padding between the section title and the body. 
You currently have an Inset of 10Dp and a UniformPadding of 16Dp.

Sirmorrison reviewed (dismissed):
<img width="798" alt="image" src="https://user-images.githubusercontent.com/27733432/116503455-0bc27000-a8ae-11eb-9d4d-ebef21541a5d.png">

The restore button is supposed to grayed out and inactive when the inputs are still empty.
<img width="413" alt="image" src="https://user-images.githubusercontent.com/27733432/116503363-c3a34d80-a8ad-11eb-9e26-01e4754300e9.png">


Also, kindly pay attention to the dimensions on the mockup so that your design aligns accordingly.

JustinBeBoy committed:
Use dimensions.go file for unit parameter
  - remove commented code
  - change dimensions between header and seed phase
  - change dimensions between card section content and card header
  - change color of restore button when inactive & active.

JustinBeBoy committed:
Directly go to restore page if from wallet page
 - Add vertical option to line.go

oshorefueled commented on ui/decredmaterial/verticalline.go:
```diff
@@ -0,0 +1,60 @@
+	paint.Fill(gtx.Ops, l.Color)
+	st.Load()
+	return layout.Dimensions{Size: line.Max}
+}
```
This can be added as part of the existing line.go as VerticalSeparator

oshorefueled commented on ui/create_restore_page.go:
```diff
@@ -629,11 +772,23 @@ func (pg *createRestore) centralize(gtx layout.Context, content layout.Widget) l
+	strength := dcrlibwallet.ShannonEntropy(password.Text()) / 4.0
+	pg.passwordStrength.Progress = float32(strength * 100)
+	pg.passwordStrength.Color = common.theme.Color.Success
+}
```
This method is also used in modal_templates.go. It can be added to the ui/util.go file and used by both modals and restore pages. This would eliminate duplication.

oshorefueled commented on ui/decredmaterial/restoreeditor.go:
```diff
@@ -0,0 +1,63 @@
+				return layout.Inset{
+					Left:  re.m10,
+					Right: re.m10,
+				}.Layout(gtx, func(gtx C) D {
```
There's no need to create another file for restore editor, it can be added to the ui/decredmaterial/editor.go file

oshorefueled commented on ui/decredmaterial/line.go:
```diff
@@ -13,12 +13,20 @@ import (
+		Height     int
+		Width      int
+		Color      color.NRGBA
+		isVetircal bool
```
typo in "isVetircal". It should be "isVertical"

oshorefueled commented on ui/decredmaterial/line.go:
```diff
@@ -13,12 +13,20 @@ import (
 )
 
+// VLine returns a vertical line widget instance
+func (t *Theme) VLine(height, width int) Line {
```
A better name for this could be "SeparatorVertical" similar to the Separator widget created in line.go

JustinBeBoy committed:
Move calculate password strength to util
 - Fix Typo
 - Move restoreeditor to editor

oshorefueled commented on ui/decredmaterial/editor.go:
```diff
@@ -51,6 +60,20 @@ func (t *Theme) EditorPassword(editor *widget.Editor, hint string) Editor {
+		Edit:       e,
+		TitleLabel: t.Body2(title),
+		LineColor:  t.Color.Gray1,
+		m10:        values.MarginPadding10,
```
You can use the dimension directly in the code as a unit.Dp value. One of the rational behind the decredmaterial package was to have it independent of godcr code such that it can easily be pulled out and used by any other project.

JustinBeBoy committed:
Using dimension directly as Dp value
